File: prep/coreprep/sorting_and_searching/BinarySearch.java
import java.util.Arrays;
import java.util.Scanner;

public class BinarySearch {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      int q = in.nextInt();
      int[] a = new int[n];

      for (int i = 0; i < n; i++) {
        a[i] = in.nextInt();
      }

      display(a, n);
      Arrays.sort(a);
      display(a, n);

      int elem, idx;
      for (int i = 0; i < q; i++) {
        elem = in.nextInt();

        System.out.println(binarySearch(a, 0, n - 1, elem));
        System.out.println(binarySearchRec(a, 0, n - 1, elem));
      }
    }
  }

  // O(logN)
  private static int binarySearch(int[] a, int low, int high, int elem) {
    int mid = -1;

    while (low <= high) {
      mid = low + (high - low) / 2;

      if (a[mid] < elem) {
        low = mid + 1;
      } else if (a[mid] > elem) {
        high = mid - 1;
      } else {
        return mid;
      }
    }

    return -1;
  }

  // O(logN)
  private static int binarySearchRec(int[] a, int low, int high, int elem) {
    if (low > high) {
      return -1;
    }

    int mid = low + (high - low) / 2;

    if (a[mid] < elem) {
      return binarySearch(a, mid + 1, high, elem);
    } else if (a[mid] > elem) {
      return binarySearch(a, low, mid - 1, elem);
    }

    return mid;
  }

  private static void display(int[] a, int n) {
    for (int i = 0; i < n; i++) {
      System.out.printf("%d ", a[i]);
    }
    System.out.println();
  }
}

File: lc/coreprep/topics/recursion/binary_search.py
def binary_search_iter(a, low, high, e):
    while low <= high:
        mid = (low + high) // 2

        if a[mid] == e:
            return mid
        elif a[mid] < e:
            low = mid + 1
        else:
            high = mid - 1
    return -1

def binary_search(a, low, high, e):
    if low > high:
        return -1

    mid = (low + high) // 2

    if a[mid] == e:
        return mid
    elif a[mid] < e:
        return binary_search(a, mid + 1, high, e)
    else:
        return binary_search(a, low, mid - 1, e)

def main():
    while True:
        try:
            (n, e) = map(int, input().split())
            a = list(map(int, input().split()))
            a.sort()
            print(f"sorted a = {a}, e = {e}")
            print(f"{binary_search(a, 0, n - 1, e)}, {binary_search_iter(a, 0, n - 1, e)}")
        except EOFError:
            break

if __name__ == "__main__":
    main()